src/lazy_es.rs

```diff
@@ -173,20 +173,24 @@ impl<'p, T: 'p> LazyEs<'p, T> {
         partials: Vec<Self>,
         max_batch_size: usize,
     ) -> Vec<T> {
-        // `partials` needs to be wrapped into a RefCell because the closure `make_progress` will
-        // return a future containing a mutable reference to it. Hence we need to ensure at runtime
-        // that this closure won't be called twice without consuming the future first.
-        let partials = RefCell::new(partials);
+        // `partials` needs to be wrapped with a `Mutex` (would be a `RefCell` in a single threaded
+        // context) because the closure `make_progress` will return a future containing a mutable
+        // reference to it. Hence we need to ensure at runtime that this closure won't be called
+        // twice without consuming the future first.
+        let partials = Mutex::new(partials);
 
         let make_progress = || async {
-            let mut partials = partials.borrow_mut();
+            let mut partials = partials
+                .try_lock()
+                .expect("`make_progress` was called concurrently");
+
             Self::batch_make_progress(es, partials.as_mut(), max_batch_size).await
         };
 
         // Don't stop while some progress has been made during the loop condition.
-        while make_progress()
+        while with_backoff(make_progress, BACKOFF_RETRIES, BACKOFF_DELAY)
```

Comment:
Turns out I had accidentally removed the main interest of previously merged PR :facepalm: 
https://github.com/Qwant/fafnir/pull/84/commits/c8faba8671cbe8059eb745915a2e6cf43b21104a#diff-288132daf6a7139a495d15c861f068e916e6571326dc52fd258fc8d66ca7c9f2L187